runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -430,6 +452,15 @@ public Timer timer(String timerId) {
     final DoFnContext<InputT, OutputT> context;
     final WindowedValue<InputT> windowedValue;
 
+    /**
+     * The state namespace for this context; lazily initialized.
+     *
+     * <p>Any call to {@link #getNamespace()} when more than one window is present will crash; this
+     * represents a bug in the runner or the {@link DoFnSignature}, since values must be in exactly
+     * one window when state or timers are relevant.
```

Comment:
You should comment that this field should not be used directly and go through `getNamespace`

If style permits, this would be great to have directly above `getNamespace` to marry the idea that it's just a cached value, and shouldn't be considered as an actual field.